Uchu.World/Systems/Match/MatchInstance.cs

```diff
@@ -0,0 +1,326 @@
+        
+        /// <summary>
+        /// Players that are required to start the match.
+        /// </summary>
+        private int _requiredPlayers;
+        
+        /// <summary>
+        /// Max players in the match.
+        /// </summary>
+        private int _maxPlayers;
+        
+        /// <summary>
+        /// Players currently in the match.
+        /// </summary>
+        private List<Player> _players = new List<Player>();
+        
+        /// <summary>
+        /// Players currently ready for the match.
+        /// </summary>
+        private List<Player> _readyPlayers = new List<Player>();
+        
+        /// <summary>
+        /// Players that were sent the time.
+        /// Sending the time multiple times causes
+        /// problems on the client.
+        /// </summary>
+        private List<Player> _playersSentTime = new List<Player>();
+
+        /// <summary>
+        /// Timer for starting the round.
+        /// </summary>
+        private Timer _countdown = new Timer()
+        {
+            AutoReset = false,
+        };
+
+        /// <summary>
+        /// Stopwatch for the elapsed time.
+        /// </summary>
+        private Stopwatch _countdownStopwatch = new Stopwatch();
+        
+        /// <summary>
+        /// Creates an instance of a match.
+        /// </summary>
+        /// <param name="type">Activity id of the match.</param>
+        public MatchInstance(int type, Zone zone)
+        {
+            // Load the match data.
+            var matchData = ClientCache.GetTable<Activities>().First(activity => activity.ActivityID == type);
+            var matchZoneId =  matchData.InstanceMapID ?? 0;
+            var matchCurrencyLot = matchData.OptionalCostLOT;
+            var matchCurrencyCount = matchData.OptionalCostCount;
+            _waitTime = matchData.WaitTime ?? 60000;
+            _playWaitTime = matchData.StartDelay ?? 5000;
+            _requiredPlayers = (matchData.MinTeams ?? 1) * (matchData.MinTeamSize ?? 1);
+            _maxPlayers = (matchData.MaxTeams ?? 1) * (matchData.MaxTeamSize ?? 1);
+            
+            // Connect the timer ending.
+            _countdown.Elapsed += async (sender, args) =>
+            {
+                // Remove the round from the provisioner.
+                TimeEnded.Invoke();
+                
+                // Take the optional currency from the players.
+                if (matchCurrencyLot.HasValue && matchCurrencyCount.HasValue)
+                {
+                    foreach (var player in _players)
+                    {
+                        if (!player.TryGetComponent<InventoryManagerComponent>(out var inventoryManager)) continue;
+                        await inventoryManager.RemoveLotAsync(matchCurrencyLot.Value,(uint) matchCurrencyCount.Value);
+                    }
+                }
+                
+                // Allocate the new zone.
+                var uchuServer = zone.UchuServer;
```

Comment:
I think it'd be nice to have this part abstracted away, there's similar logic in testmap and the launchpads I believe?